tlvc/arch/i86/boot/setup.S

```diff
@@ -976,10 +989,8 @@ hma_string:
 	.ascii	"\nhma=kernel"		// size = 11
 
 // check whether HMA can be enabled, return NC if so with A20 enabled
+// Assumes DS=INITSEG
 checkhma:
-	push	%ds
-	mov	$INITSEG,%ax
-	mov	%ax,%ds
 	mov	xms_kbytes,%ax		//. must have xms size >= 64k
 	cmp	$64,%ax
 	jl	1f
```

Comment:
As mentioned in the main comment section, don't XT-class systems set xms_bytes to 0 from the INT 15/AX=8800 call? If so, that would preclude the need for checking cpu_type when XMS memory is not present.

If on the other hand, if this check is needed for 286 systems with HMA boot hanging because XMS is available and A20 gate works, but still the HMA area is not addressable  for some reason, then yes, cpu_type == 286 will have to be checked for here. Then our additional fix of checking for arch_cpu != 6 (80286) before calling check_unreal_mode should stop further hang in xms_init.

Am I on the right track here?